albatross/models/least_squares.h

```diff
@@ -62,36 +56,56 @@ class LeastSquaresRegression
     for (int i = 0; i < m; i++) {
       A.row(i) = features[static_cast<std::size_t>(i)];
     }
-    // Solve for the coefficients using the QR decomposition.
-    LeastSquaresFit model_fit = {least_squares_solver(A, targets.mean)};
+
+    FitType model_fit = {least_squares_solver(A, targets.mean)};
     return model_fit;
   }
 
-protected:
-  Eigen::VectorXd
-  predict_mean_(const std::vector<Eigen::VectorXd> &features) const override {
+  template <typename FeatureType,
+            typename std::enable_if<
+                   has_valid_fit<ImplType, FeatureType>::value,
+                int>::type = 0>
+  FitType fit(const std::vector<FeatureType> &features,
+                    const MarginalDistribution &targets) const {
+    return impl().fit(features, targets);
+  }
+
+  JointDistribution predict(const std::vector<Eigen::VectorXd> &features,
+                            const FitType &least_squares_fit,
+                            PredictTypeIdentity<JointDistribution> &&) const {
     std::size_t n = features.size();
     Eigen::VectorXd mean(n);
     for (std::size_t i = 0; i < n; i++) {
       mean(static_cast<Eigen::Index>(i)) =
-          features[i].dot(this->model_fit_.coefs);
+          features[i].dot(least_squares_fit.coefs);
     }
-    return mean;
+    return JointDistribution(mean);
   }
 
-  JointDistribution
-  predict_(const std::vector<Eigen::VectorXd> &features) const override {
-    return JointDistribution(predict_mean_(features));
+  template <typename FeatureType, typename FitType, typename PredictType,
```

Comment:
These heavily templated methods "forward" on any `predict` methods defined in `ImplType` to this base class.  This is required since `ModelBase<LeastSquares<ImplType>>` can't actually inspect `ImplType` directly.